Products/CMFCore/indexing.py

```diff
@@ -0,0 +1,330 @@
+        to work around this a proxy object is used, which can provide the
+        original path;  of course, access to other attributes must still be
+        possible, since alternate indexers (i.e. solr etc) might use another
+        unique key, usually the object's uid;  also the inheritence tree
+        must match """
+    if getattr(aq_base(obj), 'getPhysicalPath', None) is None:
+        return obj
+
+    class PathWrapper(obj.__class__):
+
+        def __init__(self):
+            self.__dict__.update(dict(
+                context = obj,
+                path = obj.getPhysicalPath(),
+                REQUEST = getattr(obj, 'REQUEST', None)))
+
+        def __getattr__(self, name):
+            return getattr(aq_inner(self.context), name)
+
+        def __hash__(self):
+            return hash(self.context)   # make the wrapper transparent...
+
+        def getPhysicalPath(self):
+            return self.path
+
+    return PathWrapper().__of__(aq_parent(obj))
+
+
+@implementer(IIndexQueue)
+class IndexQueue(local):
+
+    def __init__(self):
+        self.queue = []
+        self.tmhook = None
+
+    def hook(self):
+        """ register a hook into the transaction machinery if that hasn't
+            already been done;  this is to make sure the queue's processing
+            method gets called back just before the transaction is about to
+            be committed """
+        if self.tmhook is None:
+            self.tmhook = QueueTM(self).register
+        self.tmhook()
+
+    def index(self, obj, attributes=None):
+        self.queue.append((INDEX, obj, attributes, None))
+        self.hook()
+
+    def reindex(self, obj, attributes=None, update_metadata=1):
+        self.queue.append((REINDEX, obj, attributes, update_metadata))
+        self.hook()
+
+    def unindex(self, obj):
+        self.queue.append((UNINDEX, wrap(obj), None, None))
+        self.hook()
+
+    def setHook(self, hook):
+        self.tmhook = hook
+
+    def getState(self):
+        return list(self.queue)     # better return a copy... :)
+
+    def setState(self, state):
+        self.queue = state
+
+    def length(self):
+        """ return number of currently queued items;  please note that
+            we cannot use `__len__` here as this will cause test failures
+            due to the way objects are compared """
+        return len(self.queue)
+
+    def optimize(self):
+        res = {}
+        for iop, obj, iattr, imetadata in self.getState():
+            oid = hash(obj)
+            func = getattr(obj, 'getPhysicalPath', None)
+            if callable(func):
+                oid = oid, func()
+            op, dummy, attr, metadata = res.get(
+                oid,
+                (0, obj, iattr, imetadata)
+            )
+            # If we are going to delete an item that was added in this
+            # transaction, ignore it
+            if op == INDEX and iop == UNINDEX:
+                del res[oid]
```

Comment:
ISTM that the state machine is more complex than you represent here.  E.g.,  `iop == UNINDEX` should cancel reindexing as well as indexing, and should then `continue`.